File: LA4-22BCT0046/LA4-22BCT0046.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "LA4-22BCT0046.h"

// Function to write the CSV header based on the sensor type
void write_csv_header(FILE *file, const char *type) {
    if (strcmp(type, "mos") == 0) {
        // Header for MOS sensor data
        fprintf(file, "Temperature,Resistance\n");
    } else if (strcmp(type, "ph") == 0) {
        // Header for pH sensor data
        fprintf(file, "pH,Potential\n");
    } else if (strcmp(type, "airbag") == 0) {
        // Header for airbag sensor data
        fprintf(file, "Temperature,Pressure,Force\n");
    }
}

int main() {
    char sensor_type[20]; // Variable to store the sensor type input by the user

    // Prompt the user to enter the sensor type
    printf("Enter sensor type (mos/ph/airbag): ");
    scanf("%s", sensor_type);

    // Open a CSV file for writing
    FILE *file = fopen("LA4-22BCT0046.csv", "w");
    if (!file) {
        // Handle file creation error
        perror("Unable to create CSV");
        return 1;
    }

    // Write the appropriate CSV header based on the sensor type
    write_csv_header(file, sensor_type);

    // Check the sensor type and process accordingly
    if (strcmp(sensor_type, "mos") == 0) {
        // Initialize MOS sensor parameters
        MOS_Sensor mos = {500, 0.5, 300}; // Ra, Ea, T

        // Loop through temperature values and calculate resistance
        for (int t = 300; t <= 500; t += 20) {
            mos.T = t; // Update temperature
            double Rg = calculate_mos_resistance(mos); // Calculate resistance
            fprintf(file, "%d,%.4e\n", t, Rg); // Write data to CSV
        }

    } else if (strcmp(sensor_type, "ph") == 0) {
        // Initialize pH sensor parameters
        PH_Sensor ph = {0.7, 7}; // E0, pH

        // Loop through pH values and calculate potential
        for (double pH_val = 4.0; pH_val <= 10.0; pH_val += 1.0) {
            ph.pH = pH_val; // Update pH value
            double voltage = calculate_potential(ph); // Calculate potential
            fprintf(file, "%.1f,%.4f\n", pH_val, voltage); // Write data to CSV
        }

    } else if (strcmp(sensor_type, "airbag") == 0) {
        // Initialize airbag sensor parameters
        Airbag airbag = {0.05, 0.06, 300, 0.5}; // n, V, T, area

        // Loop through temperature values and calculate pressure and force
        for (int T = 300; T <= 600; T += 30) {
            airbag.T = T; // Update temperature
            double pressure = calculate_pressure(airbag); // Calculate pressure
            double force = calculate_force(pressure, airbag.surface_area); // Calculate force
            fprintf(file, "%d,%.2f,%.2f\n", T, pressure, force); // Write data to CSV
        }

    } else {
        // Handle invalid sensor type
        printf("Invalid sensor type.\n");
        fclose(file); // Close the file
        return 1;
    }

    // Close the file after writing data
    fclose(file);
    printf("Data saved to LA4-22BCT0046.csv\n");
    return 0; // Exit the program successfully
}

File: LA4-22BCT0046/LA4-22BCT0046.py
import pandas as pd
import matplotlib.pyplot as plt

# Set dark theme for the plots
plt.style.use('dark_background')

# Prompt the user to input the sensor type they want to visualize
sensor = input("Enter sensor type to plot (mos/ph/airbag): ").strip().lower()

# Load the sensor data from a CSV file into a pandas DataFrame
df = pd.read_csv("LA4-22BCT0046.csv")

# Set the main title for the visualization
plt.suptitle("Sensor Data Visualization - 22BCT0046", fontsize=16, fontweight='bold')

# Check the sensor type and plot the corresponding data
if sensor == "mos":
    # Plot Resistance vs Temperature for MOS sensor
    plt.plot(df["Temperature"], df["Resistance"], color="#00FFFF", linewidth=2, marker='o', label="Resistance")
    plt.title("MOS Sensor: Resistance vs Temperature", fontsize=14)
    plt.xlabel("Temperature (K)", fontsize=12)
    plt.ylabel("Resistance (Ohms)", fontsize=12)
elif sensor == "ph":
    # Plot Potential vs pH for Potentiometric sensor
    plt.plot(df["pH"], df["Potential"], color="#FF69B4", linewidth=2, marker='s', label="Potential")
    plt.title("Potentiometric Sensor: Potential vs pH", fontsize=14)
    plt.xlabel("pH Value", fontsize=12)
    plt.ylabel("Potential (V)", fontsize=12)
elif sensor == "airbag":
    # Plot Force and Pressure vs Temperature for Airbag sensor
    plt.plot(df["Temperature"], df["Force"], color="#00FF7F", linewidth=2, marker='^', label="Force")
    plt.plot(df["Temperature"], df["Pressure"], color="#FFD700", linewidth=2, marker='x', label="Pressure")
    plt.title("Airbag Sensor: Force & Pressure vs Temperature", fontsize=14)
    plt.xlabel("Temperature (K)", fontsize=12)
    plt.ylabel("Force / Pressure", fontsize=12)
    plt.legend()  # Add a legend to distinguish between Force and Pressure
else:
    # Handle invalid sensor type input
    print("Invalid sensor type.")
    exit()

# Add a grid to the plot for better readability
plt.grid(True, linestyle="--", alpha=0.5)

# Adjust the layout to prevent overlapping of elements
plt.tight_layout()

# Display the plot
plt.show()
